ETH-Merge-Bots/Changes-in-Block-Difficulty/src/agent.ts

```diff
@@ -0,0 +1,71 @@
+import { Finding, HandleBlock, BlockEvent, getEthersProvider, ethers } from "forta-agent";
+import { NUMBER_OF_BLOCKS_TO_CHECK, FINDING_THRESHOLD, createFinding, TERMINAL_TOTAL_DIFFICULTY } from "./utils";
+import BigNumber from "bignumber.js";
+
+BigNumber.set({ DECIMAL_PLACES: 2 });
+
+export const provideHandleBlock = (
+  init: boolean,
+  provider: ethers.providers.Provider,
+  numberOfBlocksToCheck: number,
+  threshold: BigNumber
+): HandleBlock => {
+  let blockDifficulties: BigNumber[] = new Array<BigNumber>();
+  let blockDifficultyMovingAverage: BigNumber;
+  let blockDifficultiesSum: BigNumber = new BigNumber(0);
+
+  return async (blockEvent: BlockEvent) => {
+    const findings: Finding[] = [];
+
+    // check if TTD has been reached
+    if (new BigNumber(blockEvent.block.totalDifficulty).gte(TERMINAL_TOTAL_DIFFICULTY)) return findings;
+
+    // executed on the first run
+    if (!init) {
+      init = true;
+      //populate the blockDifficulties array in ascending block order
+      for (let i = 0; i <= numberOfBlocksToCheck - 1; i++) {
+        const blockDifficulty = new BigNumber(
+          (await provider.getBlock(blockEvent.blockNumber - numberOfBlocksToCheck + i))._difficulty.toString()
+        );
+        blockDifficulties.push(blockDifficulty);
+      }
+    }
+
+    //remove the oldest block difficulty
+    blockDifficulties.shift();
+    //add new block difficulty
+    const currentDifficulty: BigNumber = new BigNumber(blockEvent.block.difficulty);
+    blockDifficulties.push(currentDifficulty);
+    //calculate the new moving average
+    for (let i = 0; i <= blockDifficulties.length - 1; i++) {
+      blockDifficultiesSum = blockDifficultiesSum.plus(blockDifficulties[i]);
+    }
```

Comment:
this is okay and fast but this can also be written with `reduce` function. just fyi.

```suggestion
    blockDifficultiesSum = blockDifficulties.reduce(
      (accumulator: BigNumber, curr: BigNumber) => accumulator.plus(curr),
      new BigNumber("0")
    );
```

Reply:
That's way better, and I can omit the `new BigNumber("0")` as it's initialized above, right?

Reply:
yes that's the starting initial sum, you can keep it or remove it but it's generally kept to clearly show that the initial value starts from 0

Reply:
and when you do this, some tests fail. so you might wanna check them too

Reply:
Can you `git pull` and check again? They seem to be working fine for me

Reply:
oh okay now it worked 👍 btw now you can remove `blockDifficultiesSum = new BigNumber(0);` it's not necessary anymore